### PR TITLE
vecdb caps support

### PR DESCRIPTION
**Pull request by valaises**

### CODE REVIEW

File: src/vecdb/vectorizer_service.rs [hunk 1]
```diff
@@ -108,41 +114,61 @@ async fn vectorize_thread(
         drop(vecdb_handler);
         info!("Retrieving embeddings for {} chunks", split_data_filtered.len());
 
-        let join_handles: Vec<_> = split_data_filtered.iter().map(
-            |x| get_embedding(x.window_text.clone(), &embedding_model_name, api_key.clone())
-        ).collect();
-        status.lock().await.requests_made_since_start += join_handles.len();
```
Comment (JegernOUTT) [inline src/vecdb/vectorizer_service.rs:114]: Have you moved it somewhere?

**Review by JegernOUTT — CHANGES_REQUESTED** [review]

File: src/vecdb/vecdb.rs [hunk 2]
```diff
@@ -28,22 +84,32 @@ impl VecDb {
...
-            vecdb_handler.clone(), cooldown_secs, splitter_window_size, splitter_soft_limit,
-            embedding_model_name.clone(), cmdline.api_key.clone(),
+        let vecdb_handler = Arc::new(AMutex::new(handler));
+        let retriever_service = Arc::new(AMutex::new(FileVectorizerService::new(
+            vecdb_handler.clone(),
+            cooldown_secs,
+            splitter_window_size,
+            splitter_soft_limit,
+            embedding_model_name.clone(),
+            cmdline.api_key.clone(),
```
Comment (JegernOUTT) [inline src/vecdb/vecdb.rs:101]: Let's rethink using of the internal api_key 
Maybe we should add extra args like `hf_api_key`, `openai_api_key` and use them instead

**Review by JegernOUTT — DISMISSED** [review]

File: src/vecdb/vectorizer_service.rs (hunk 1 above)
Comment (valaises) [reply src/vecdb/vectorizer_service.rs:114]: no :) it just expanded
```rust
        let join_handles: Vec<_> = split_data_filtered.into_iter().map(|x| {
            let embedding_model_name_clone = embedding_model_name.clone();
            let api_key_clone = api_key.clone();
            let endpoint_style_clone = endpoint_style.clone();
            let url_clone = url.clone();

            let semaphore_clone = Arc::clone(&semaphore);
            tokio::spawn(async move {
                let _permit = match semaphore_clone.acquire().await {
                    Ok(permit) => permit,
                    Err(_) => {
                        return None;
                    }
                };

                let result = get_embedding(
                    &endpoint_style_clone,
                    &embedding_model_name_clone,
                    &url_clone,
                    x.window_text.clone(),
                    &api_key_clone,
                ).await;

                drop(_permit);
                Some((x, result))
            })
        }).collect();
```

File: src/vecdb/vecdb.rs (hunk 2 above)
Comment (valaises) [reply src/vecdb/vecdb.rs:101]: @olegklimov do you think we should?

File: src/vecdb/req_client.rs [hunk 3]
```diff
@@ -5,42 +5,69 @@ use tokio::task::JoinHandle;
...
+        )
+    }
+    else {
+        panic!("Invalid endpoint style: {}", endpoint_style);
     }
 }
 
-pub fn get_embedding(
+
+fn get_embedding_hf_style(
```
Comment (olegklimov) [inline src/vecdb/req_client.rs:51]: to forward_to_hf_endpoint.rs

File: src/vecdb/req_client.rs [hunk 4]
```diff
@@ -61,14 +88,82 @@ pub fn get_embedding(
...
+                Err(err) => return Err(format!("Failed to send a request: {:?}", err)),
+            }
+        }
+
+        Err("Exceeded maximum attempts to reach the server".to_string())
+    })
+}
+
+
+fn get_embedding_openai_style(
```
Comment (olegklimov) [inline src/vecdb/req_client.rs:100]: to forward_to_openai_endpoint.rs

File: src/vecdb/req_client.rs [hunk 5]
```diff
@@ -5,42 +5,69 @@ use tokio::task::JoinHandle;
 use tokio::time::sleep;
```
Comment (olegklimov) [inline src/vecdb/req_client.rs:6]: fetch_embedding.rs in root near restream.rs

File: src/caps.rs [hunk 6]
```diff
@@ -44,6 +44,10 @@ pub struct CodeAssistantCaps {
     #[serde(default)]
     pub code_chat_models: HashMap<String, ModelRecord>,
     pub code_chat_default_model: String,
+    #[serde(default)]
+    pub embeddings_default_model: String,
+    #[serde(default)]
+    pub embeddings_endpoint_template: String,
```
Comment (olegklimov) [inline src/caps.rs:50]: endpoint_embedding_template
endpoint_embedding_style

File: src/caps.rs [hunk 7]
```diff
@@ -344,13 +348,28 @@ pub async fn load_caps(
...
+    let mut embeddings_endpoint_template = r1.embeddings_endpoint_template.clone();
+    if r1.cloud_name == "Hugging Face" {
+        embeddings_endpoint_template = format!("{}/models/{}", "https://api-inference.huggingface.co".to_string(), r1.embeddings_default_model);
+        Ok(embeddings_endpoint_template)
+    } else {
+        embeddings_endpoint_template = relative_to_full_url(&caps_url, &embeddings_endpoint_template)?;
+        Ok(embeddings_endpoint_template)
+    }
+}
+
```
Comment (olegklimov) [inline src/caps.rs:372]: ???

**Review by reymondzzzz — APPROVED** [review]

File: src/global_context.rs [hunk 8]
```diff
@@ -42,6 +42,10 @@ pub struct CommandLine {
     pub insecure: bool,
     #[structopt(long, help="Whether to use a vector database")]
     pub vecdb: bool,
+    #[structopt(long, default_value="Refact", help="vecdb provider")]
+    pub vecdb_provider: String,
+    #[structopt(long, default_value="", help="vecdb api key")]
+    pub vecdb_api_key: String,
     #[structopt(long, short = "f", default_value = "", help = "The path to jsonl file which contains filtered source files")]
```
Comment (olegklimov) [inline src/global_context.rs:49]: embedding, not vecdb

File: src/main.rs [hunk 9]
```diff
@@ -57,16 +60,33 @@ async fn main() {
             info!("{:>20} {}", k, v);
         }
     }
+    let caps_mb = global_context::try_load_caps_quickly_if_not_present(gcx.clone()).await;
```
Comment (olegklimov) [inline src/main.rs:63]: bad idea

File: src/main.rs [hunk 10]
```diff
@@ -57,16 +60,33 @@ async fn main() {
...
     let lsp_task = spawn_lsp_task(gcx.clone(), cmdline.clone()).await;  // execution stays inside if stdin-stdout
     if lsp_task.is_some() {
         background_tasks.push_back(lsp_task.unwrap())
     }
-    background_tasks.extend(match *gcx.read().await.vec_db.lock().await {
-        Some(ref db) => db.start_background_tasks().await,
-        None => vec![]
-    });
+    {
+        let vecdb_mb = vecdb::vecdb::create_vecdb_if_caps_present(gcx.clone()).await;
```
Comment (olegklimov) [inline src/main.rs:81]: bad idea, should only accept global_context, work with caps dynamically changing

File: src/vecdb/vecdb.rs [hunk 11]
```diff
@@ -2,20 +2,100 @@ use std::path::PathBuf;
...
     embedding_model_name: String,
     cmdline: crate::global_context::CommandLine,
+    endpoint_embedding: String,
+    provider_embedding: String,
+}
+
+
+fn resolve_endpoint_embeddings_url(
+    endpoint_embeddings_template: &String,
+    vecdb_provider: &String,
```
Comment (olegklimov) [inline src/vecdb/vecdb.rs:30]: style

File: src/vecdb/vecdb.rs [hunk 12]
```diff
@@ -2,20 +2,100 @@ use std::path::PathBuf;
...
+    embeddings_default_model: &String,
+) -> String {
+    return if vecdb_provider == "hf" {
+        format!("{}/models/{}", "https://api-inference.huggingface.co".to_string(), embeddings_default_model)
```
Comment (olegklimov) [inline src/vecdb/vecdb.rs:34]: hardcoded URL

File: src/fetch_embedding.rs [hunk 13]
```diff
@@ -1,74 +1,45 @@
...
-    #[cfg(not(test))]
-    {
-        // In a regular run, return the actual URL
-        "https://api-inference.huggingface.co".to_string()
-    }
-}
 
 pub fn get_embedding(
-    text: String,
+    provider_embedding: &String,
```
Comment (olegklimov) [inline src/fetch_embedding.rs:12]: template

Comment (olegklimov) [reply src/fetch_embedding.rs:12]: style

File: src/fetch_embedding.rs [hunk 14]
```diff
@@ -1,74 +1,45 @@
...
     model_name: &String,
-    api_key: String,
+    url: &String,
```
Comment (olegklimov) [inline src/fetch_embedding.rs:14]: template

File: src/forward_to_hf_endpoint.rs [hunk 15]
```diff
@@ -85,3 +89,58 @@ pub async fn forward_to_hf_style_endpoint_streaming(
...
+    let client = reqwest::Client::new();
+    let payload = EmbeddingsPayloadHF { inputs: text };
+
+    let url_clone = url.clone();
+    let api_key_clone = api_key.clone();
+
+    tokio::spawn(async move {
+        let mut attempts = 0;
+
+        while attempts < max_attempts {
```
Comment (olegklimov) [inline src/forward_to_hf_endpoint.rs:116]: don't need attempts, should be self-fixing logic in the vecdb

File: src/fetch_embedding.rs [hunk 16]
```diff
@@ -1,74 +1,40 @@
...
-    #[cfg(not(test))]
-    {
-        // In a regular run, return the actual URL
-        "https://api-inference.huggingface.co".to_string()
-    }
-}
 
 pub fn get_embedding(
-    text: String,
+    cloud_name: &String,
```
Comment (olegklimov) [inline src/fetch_embedding.rs:10]: Cloud name is not a thing

File: src/forward_to_hf_endpoint.rs [hunk 17]
```diff
@@ -85,3 +89,45 @@ pub async fn forward_to_hf_style_endpoint_streaming(
...
+    pub inputs: String,
+}
+
+
+pub fn get_embedding_hf_style(
+    text: String,
+    endpoint_template: &String,
+    model_name: &String,
+    api_key: &String,
+) -> JoinHandle<Result<Vec<f32>, String>> {
```
Comment (olegklimov) [inline src/forward_to_hf_endpoint.rs:105]: Why not just async?

File: src/forward_to_openai_endpoint.rs [hunk 18]
```diff
@@ -103,3 +107,60 @@ fn _passthrough_messages_to_json(
...
+    pub model: String,
+}
+
+
+pub fn get_embedding_openai_style(
+    text: String,
+    endpoint_template: &String,
+    model_name: &String,
+    api_key: &String,
+) -> JoinHandle<Result<Vec<f32>, String>> {
```
Comment (olegklimov) [inline src/forward_to_openai_endpoint.rs:124]: async?

File: src/global_context.rs [hunk 19]
```diff
@@ -42,7 +42,7 @@ pub struct CommandLine {
     pub insecure: bool,
     #[structopt(long, help="Whether to use a vector database")]
     pub vecdb: bool,
-    #[structopt(long, short = "f", default_value = "", help = "The path to jsonl file which contains filtered source files")]
+    #[structopt(long, default_value="Refact", help="vecdb provider")]
```
Comment (olegklimov) [inline src/global_context.rs:45]: jsonl file, not vecdb provider

File: src/vecdb/vecdb.rs [hunk 20]
```diff
@@ -2,52 +2,187 @@ use std::path::PathBuf;
...
+            default_embeddings_model, cloud_name, endpoint_embeddings_template, cmdline, cache_dir
+        ).await;
+
+        if vecdb_mb.is_none() {
+            info!("vecd fetch failed: vecdb_mb.is_none()");
+            continue;
+        }
+        gcx_locked.vec_db = Arc::new(AMutex::new(vecdb_mb));
+        info!("VECDB is successfully fetched");
+        vecdb_fetched = true;
```
Comment (olegklimov) [inline src/vecdb/vecdb.rs:99]: Fetched what?..

File: src/vecdb/vectorizer_service.rs [hunk 21]
```diff
@@ -170,8 +203,11 @@ impl FileVectorizerService {
         cooldown_secs: u64,
         splitter_window_size: usize,
         splitter_soft_limit: usize,
-        embedding_model_name: String,
+
+        model_name: String,
         api_key: String,
+        cloud_name: String,
+        endpoint_template: String,
```
Comment (olegklimov) [inline src/vecdb/vectorizer_service.rs:210]: global_context? 

To quit if caps changes significantly.

cloud_name is not a thing.

File: src/vecdb/vecdb.rs [hunk 22]
```diff
@@ -2,52 +2,187 @@ use std::path::PathBuf;
...
-    retriever_service: Arc<Mutex<FileVectorizerService>>,
-    embedding_model_name: String,
-    cmdline: crate::global_context::CommandLine,
+    retriever_service: Arc<AMutex<FileVectorizerService>>,
+    cmdline: CommandLine,
+
+    model_name: String,
+    api_key: String,
+    cloud_name: String,
+    endpoint_template: String,
```
Comment (olegklimov) [inline src/vecdb/vecdb.rs:28]: api_key is the only thing that doesn't change.

cloud_name should not exist at all.

model_name, endpoint_template -- changes with caps. A pointer to caps is better.

**Pull request merged**